### PR TITLE
Heightmap reference tracking implementation

### LINKED ISSUE #94: Heightmap unit tests

### PR DESCRIPTION
**Pull request by NotStirred**
`SurfaceTrackerSection` was split into a base class `SurfaceTrackerNode`, and two derived classes: `SurfaceTrackerBranch`, `SurfaceTrackerLeaf`, which splits their implementation making it far simpler.

All of the `Light` specific `SurfaceTrackerSection` stuff is now gone, leaving only the `LightSurfaceTrackerWrapper` and the light cube implementation remaining.

`SurfaceTrackerBranch`es have a `requiredChildren` value, which for scale 1 branches
is the number of their children (leaves) with a cube loaded. Any node with
`requiredChildren` will be required by its parent.
If the `requiredChildren` is 0, all children can be unloaded.

### CODE REVIEW

**Comment by NotStirred** [comment]
More tests to come, once I get around to writing them

**Comment by NotStirred** [comment]
I'd like to write some debug output for runtime that verifies the state of the heightmap tree for each column

Once I have, I'll be much more confident to merge

File: config/checkstyle/suppressions.xml [hunk 1]
```diff
@@ -21,13 +21,13 @@
...
     <!-- TODO: These probably need refactoring -->
-    <suppress checks="CyclomaticComplexity" files="CubicLevelLoadingScreen.java|CubeSerializer.java|DebugVisualization.java|LightSurfaceTrackerSection.java"/>
+    <suppress checks="CyclomaticComplexity" files="CubicLevelLoadingScreen.java|CubeSerializer.java|DebugVisualization.java|SurfaceTrackerNodesTest.java"/>
     <!-- TODO: DebugVulkan doesn't work yet anyway, and this requires some refactoring -->
     <suppress checks="NPathComplexity" files="DebugVulkan.java"/>
     <!-- Mixins don't count for now until ASM is done -->
     <suppress checks="NPathComplexity" files=".*Mixin.*\.java"/>
     <!-- TODO These need refactoring -->
-    <suppress checks="NPathComplexity" files="LevelCube.java|ProtoCube.java|CubeSerializer.java"/>
+    <suppress checks="NPathComplexity" files="LevelCube.java|ProtoCube.java|CubeSerializer.java|SurfaceTrackerNodesTest.java"/>
```
Comment (Niko-sk2x) [inline config/checkstyle/suppressions.xml]: I don't think these should be necessary for *tests*.

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/debug/common/MixinServerLevel.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                    verifyHeightmapTree(((LightHeightmapGetter) levelChunk).getServerLightHeightmap().getSurfaceTrackerSection(), ((ColumnCubeMapGetter) levelChunk).getCubeMap());
+                }
+            });
+        }
+    }
+
+    /**
+     * Rough duplicate of SurfaceTrackerNodesTest#verifyHeightmapTree, which is in the test module
+     */
+    private void verifyHeightmapTree(SurfaceTrackerBranch root, ColumnCubeMap cubeMap) {
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/debug/common/MixinServerLevel.java:84]: This seems to only check that every part of the surface tracker tree that is loaded is required, but not that all required parts are loaded, is that correct?

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/level/chunk/ProtoCube.java [hunk 3]
```diff
@@ -238,23 +241,48 @@ public void onEnteringFeaturesStatus() {
...
+        SurfaceTrackerLeaf[] lightHeightmapLeaves = this.lightHeightmaps;
+        for (int localSectionZ = 0; localSectionZ < CubeAccess.DIAMETER_IN_SECTIONS; localSectionZ++) {
+            for (int localSectionX = 0; localSectionX < CubeAccess.DIAMETER_IN_SECTIONS; localSectionX++) {
+                int i = localSectionX + localSectionZ * CubeAccess.DIAMETER_IN_SECTIONS;
+                if (lightHeightmapLeaves[i] != null) {
+                    lightHeightmapLeaves[i].cubeUnloaded(localSectionX, localSectionZ, storage);
+                    lightHeightmapLeaves[i] = null;
+                }
+            }
+        }
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/world/level/chunk/ProtoCube.java:278]: to avoid duplication like this, may be better to have a method that returns an iterable of all heightmaps if this is done in  more places.

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerNode.java [hunk 4]
```diff
@@ -176,6 +131,15 @@ protected void setDirty(int idx) {
         dirtyPositions[idx >> 6] |= 1L << idx;
     }
 
+    public void setAllDirty() {
+        assert dirtyPositions.length == 4;
+
+        dirtyPositions[0] = -1;
+        dirtyPositions[1] = -1;
+        dirtyPositions[2] = -1;
+        dirtyPositions[3] = -1;
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerNode.java:140]: Why didn't we use Arrays.fill here?

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerSectionStorage.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package io.github.opencubicchunks.cubicchunks.world.level.levelgen.heightmap.surfacetrackertree;
+
+import java.util.Arrays;
+
+import io.github.opencubicchunks.cubicchunks.world.level.levelgen.heightmap.HeightmapStorage;
+import it.unimi.dsi.fastutil.objects.Object2ReferenceMap;
+import it.unimi.dsi.fastutil.objects.Object2ReferenceOpenHashMap;
+
+public class SurfaceTrackerSectionStorage implements HeightmapStorage {
+    private Object2ReferenceMap<PackedTypeScaleScaledY, SurfaceTrackerNode> saved = new Object2ReferenceOpenHashMap<>();
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerSectionStorage.java:11]: This needs a comment explaining that this is not going to stay that way

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerSectionStorage.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    @Override
+    public SurfaceTrackerNode loadNode(SurfaceTrackerBranch parent, byte heightmapType, int scale, int scaledY) {
+        SurfaceTrackerNode removed = saved.remove(new PackedTypeScaleScaledY(heightmapType, scale, scaledY));
+        if (removed != null) {
+            removed.parent = parent;
+        }
+        return removed;
+    }
+
+    record PackedTypeScaleScaledY(byte heightmapType, int scale, int scaledY) { }
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerSectionStorage.java:34]: Maybe rename to something like `NodeLocation`?

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]
Overall seems fine, just a few small things

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerSectionStorage.java (hunk 6 above)
Comment (NotStirred) [reply src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerSectionStorage.java:34]: that's also not staying, just for the hashmap

File: config/checkstyle/suppressions.xml (hunk 1 above)
Comment (NotStirred) [reply config/checkstyle/suppressions.xml]: It fails check otherwise

Comment (Niko-sk2x) [reply config/checkstyle/suppressions.xml]: Yes... the point is to write the code so that it doesn't... that's the point of these checks.

Comment (NotStirred) [reply config/checkstyle/suppressions.xml]: cyclomatic complexity of a test is naturally going to be high, it's a single method that needs to check various things and edge-cases. I could split it into other methods but that's pointless

Comment (Niko-sk2x) [reply config/checkstyle/suppressions.xml]: There was no way to tell github that, but that was also in relation to the exception added above.
But I guess for that one, sure. I would still prefer it refactored.

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerNode.java (hunk 4 above)
Comment (NotStirred) [reply src/main/java/io/github/opencubicchunks/cubicchunks/world/level/levelgen/heightmap/surfacetrackertree/SurfaceTrackerNode.java:140]: No reason other than the possible tiny performance improvement

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/level/chunk/ProtoCube.java (hunk 3 above)
Comment (NotStirred) [reply src/main/java/io/github/opencubicchunks/cubicchunks/world/level/chunk/ProtoCube.java:278]: I assume you're referring to the normal/light heightmaps? Yeah that could be nice, though we'd need a `LIGHT` heightmap type, or return a byte map or something

Comment (NotStirred) [reply src/main/java/io/github/opencubicchunks/cubicchunks/world/level/chunk/ProtoCube.java:278]: Made a note about this in [the project](https://github.com/orgs/OpenCubicChunks/projects/1)
